Added drivers for multiple tree distance calculations

qiyunzhu opened the pull request:
Please complete the following checklist:

* [x] I have read the [contribution guidelines](https://scikit.bio/contribute.html).

* [x] I have documented all public-facing changes in the [changelog](https://github.com/scikit-bio/scikit-bio/blob/main/CHANGELOG.md).

* [ ] **This pull request includes code, documentation, or other content derived from external source(s).** If this is the case, ensure the external source's license is compatible with scikit-bio's [license](https://github.com/scikit-bio/scikit-bio/blob/main/LICENSE.txt). Include the license in the `licenses` directory and add a comment in the code giving proper attribution. Ensure any other requirements set forth by the license and/or author are satisfied.

  - **It is your responsibility to disclose** code, documentation, or other content derived from external source(s). If you have questions about whether something can be included in the project or how to give proper attribution, include those questions in your pull request and a reviewer will assist you.

* [x] This pull request does not include code, documentation, or other content derived from external source(s).

Note: [This document](https://scikit.bio/devdoc/review.html) may also be helpful to see some of the things code reviewers will be verifying when reviewing your pull request.

wasade commented on skbio/tree/_dists.py:
```diff
@@ -0,0 +1,395 @@
+    sets = [getattr(x, method)(within=y) for x, y in zip(trees, withins)]
+    lens = [len(s) for s in sets]
+
+    result = np.zeros((n_trees := len(trees), n_trees))
+    for i, j in combinations(range(n_trees), 2):
+        rf = len(sets[i] ^ sets[j])
+        if proportion:
+            total = lens[i] + lens[j]
+            rf = rf / total if total else 1.0
+        result[i, j] = result[j, i] = rf
```
It looks like this may be a re-implementation of RFD? I understand that this may be faster than to filter the trees and use `TreeNode.compare_rfd`, and suspect that's why it was done. However, it comes at the expense of maintenance burden, the potential for differing results based on the exact implementation used, and may violate the DRY principle. 

If the existing `compare_rfd` method is used, then there is a single point in the codebase to maintain for RFD, and a single block of code to consider optimizing in the future.

qiyunzhu replied:
Good point. I am still in the process of refining the code. Ideally, each algorithm is only implemented once. There are some discrepancies between two tree RF and multi-tree RF, because the former also takes care of some non-RF metrics. But I might be able to overcome this.

wasade replied:
I wonder whether a single implementation of RF that accepts `Iterable[set | frozen set]` (i.e., agnostic to `TreeNode`, etc), possibly an implementation specific for wRF, should be provided. And then this code, as well as the respective `TreeNode.compare_*` methods call that single stable implementation?

qiyunzhu replied:
I have implemented your suggestion. Specifically, in `_compare.py` there is a driver function `_topo_dists` which calls `_setdiff` or `_mapdiff` depending on whether trees are rooted.

wasade commented on skbio/tree/_dists.py:
```diff
@@ -0,0 +1,395 @@
+    sets = [frozenset(x) for x in maps]
+
+    result = np.zeros((n_trees := len(trees), n_trees))
+    for i, j in combinations(range(n_trees), 2):
```
Same comment here, it seems like there is a missed opportunity to funnel this work all through the same implementation.

wasade commented on skbio/tree/_dists.py:
```diff
@@ -0,0 +1,395 @@
+    if sample is not None:
+        shuffler = _check_shuffler(shuffler)
+
+    if pairwise:
```
There is a fair amount of common boiler plate code here that likely can be abstracted out. This branch is could probably be a function with the signature along the lines of `_pairwise(tree_function: Callable, ids: Iterable[str], **arguments) -> skbio.DistanceMatrix`?

qiyunzhu replied:
Good suggestion!

qiyunzhu replied:
Done.

wasade commented on skbio/tree/_dists.py:
```diff
@@ -0,0 +1,395 @@
+            ignore_self=True,
+        )
+        return DistanceMatrix.from_iterable(
+            trees, metric=metric, keys=ids, validate=False
```
`metric` here I _think_ would be the `repr` of the `partial` function from line 358 which is likely not what's intended

wasade replied:
FWIW, I've recently shifted to using PyRight as an LSP, alongside ruff. It would be painful to adapt in skbio because it actually uses typing within the IDE, and I think would require we move to type the project as a whole. But, it is very good at catching these types of issues, among others, so there may be long term benefit.

qiyunzhu replied:
Great thought! I guess that typing can facilitate JIT by future Python interpreters. Since we ditch Python 3.8 now, I think that Python 3.9+ has better built-in support for type hinting. One the other hand, enforcing typing will require lots of efforts though, and some may not be straightforward (therefore cannot be directly outsourced to an undergrad). SciPy and scikit-learn haven't adopted typing, but pandas has.

wasade commented on skbio/tree/_tree.py:
```diff
@@ -4306,9 +4310,9 @@ def _compare_topology(
         if not weighted:
             if symmetric:
-                result = sets1.symmetric_difference(sets2)
+                result = sets1 ^ sets2
```
Many developers won't recognize what operation is being performed here and will need to google it. Can the explicit form be retained?

qiyunzhu replied:
Makes sense! Let me add a comment after each line of code.

wasade reviewed (commented):
Added a few comments, thanks!

qiyunzhu commented on skbio/tree/_dists.py:
```diff
@@ -0,0 +1,395 @@
+    taxon_sets = [t.subset() for t in trees]
+    shared = frozenset.intersection(*taxon_sets)
+    if not shared:
+        raise ValueError("No taxon is shared across all trees.")
```
The trees will br shrinked to only shared taxa prior to the calculation of Robinson-Foulds distances. If the trees are empty, the RF distances will become 0, and the normalized RF distances will be 0 / 0 = 1. This is technically doable. I am not sure how justified it is though. Let me think...

qiyunzhu replied:
I implemented your suggestion. Specifically, if two trees share no taxon, RF will be 0 and normalized RF will be 1. In this way, any input trees will return values rather than throwing errors.

qiyunzhu commented on skbio/tree/_dists.py:
```diff
@@ -0,0 +1,395 @@
+    ids : list of str, optional
+        Unique identifiers of input trees. If omitted, will use incremental integers
+        "0", "1", "2",...
+    pairwise : bool, optional
```
Good point. I hesitated between `pairwise` and `shared_by_all`. Given your vote, let me go with the latter.

qiyunzhu commented:
@wasade Thank you for reviewing and suggesting! I think I have implemented all of the suggestions you made. Most importantly, the algorithms are now implemented only once, with all public-facing functions calling them to achieve different outputs.
